pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -221,37 +221,35 @@ def set_config(self, cfg: SimConfig) -> None:
         kraus_ops = []
         self._collapse_ops = []
         if "dephasing" in self.config.noise:
-            # Probability of phase (Z) flip:
-            # First order in prob
-            prob = self.config.dephasing_prob / 2
             n = self._size
-            if prob > 0.1 and n > 1:
-                warnings.warn(
-                    "The dephasing model is a first-order approximation in the"
-                    f" dephasing probability. p = {2*prob} is too large for "
-                    "realistic results.",
-                    stacklevel=2,
-                )
-            k = np.sqrt(prob * (1 - prob) ** (n - 1))
-
+            if self.basis_name == "all" and self.dim == 3:  # three-level system
+                prob = self.config.dephasing_prob / 3
+                if prob > 0.1 and n > 1:
+                    warnings.warn(
+                        "The dephasing model is a first-order approximation in the"
+                        f" dephasing probability. p = {3*prob} is too large for "
+                        "realistic results.",
+                        stacklevel=2,
+                    )
+                M_0 = np.sqrt(prob/(1-prob)) * qutip.qeye(3)
+                M_i = [np.sqrt(prob) * qutip.ket(str(i), 3) * qutip.bra(str(i), 3) for i in range(3)]
+                kraus_ops.append(M_0)
+                kraus_ops.append(M_i)
+            else:  # two-level system
+                prob = self.config.dephasing_prob / 2
+                if prob > 0.1 and n > 1:
+                    warnings.warn(
+                        "The dephasing model is a first-order approximation in the"
+                        f" dephasing probability. p = {2*prob} is too large for "
+                        "realistic results.",
+                        stacklevel=2,
+                    )
```

Comment:
This warning can probably be moved to after the `if-else` clause to avoid repetition